src/ws/cockpitauth.c

```diff
@@ -502,6 +509,7 @@ session_start_process (const gchar **argv,
     }
 
   child.io = fds[0];
+  child.credentials_fd = credentials_fd;
```

Comment:
I'm pretty sure that FD passing won't work for sudo + PAM modules. We need to use something like the kernel keyring instead. Do you want to merge this solution here, and then revert it later when migrating to something like the kernel keyring?

Reply:
I really want to avoid the kernel keyring. It will prevent having this working from e. g. a cockpit/ws (bastion) container. But indeed fd passing through `fork()` will obviously not work for the future case when `libpam_cockpit_cert` gets added to sudo's PAM stack, and it needs to ask cockpit-tls directly.

For that use case I thought that the future cockpit-tls needs to expose a Unix socket in e. g. `/run/cockpit/tls-socket`, restricted so that only cockpit-tls and root can access it (i. e. in particular *not* cockpit-ws). Then the PAM module can ask cockpit-tls via that Unix socket about "do you have a certificate for me for user foo" and get the memfd over that Unix socket. In that scenario it wouldn't actually have to be a memfd, as direct cockpit-tls → sudo PAM module is a trusted path. But we still need it when we inherit the memfd via cockpit-ws on normal login.

As for actually triggering that certificate, an initial idea: we let our Cockpit shell code watch a file `$XDG_RUNTIME_DIR/cockpit-cert-auth-request`. If it gets updated, the shell code triggers a new HTTPS request to some dummy cockpit URL, and cockpit-tls can acquire a fresh certificate presentation through that. The PAM module can touch that file and then ask cockpit-tls (through the Unix socket) for the certificate.

Perhaps we find a way to not have two different ways of passing the certificate, and just always use the direct cockpit PAM module → cockpit-tls unix socket path. If we have that, then we could drop memfd again.

Reply:
How will cockpit-tls know which TLS session to associate with which access from a unix socket?

The kernel keyring has sessions, which are essentially process trees. So I was thinking we could use those. When avoiding the kernel keyring, we'd need a replacement for that aspect.

Reply:
The fundamental question is how cockpit-session (or more specifically, pam_cockpit_cert) can identify itself so that cockpit-tls can give it the right cert; it doesn't have knowledge about its TLS connection and it can't ask cockpit-ws (as we don't want to trust it).

The session keyring wouldn't help with this problem, AFAICS. cockpit-tls, all cockpit-ws, and cockpit-session (before it forks and the child starts a new session) are all running in the same session. And if each session starts its own keyring, then once again you have the problem above.

This is not a big issue if we are already running in a Cockpit session, and the PAM module gets invoked through sudo -- then this knows the session's user name, and cockpit-tls can map that to a certificate. However, at cockpit login time we don't have an user name yet (as that's precisely what the PAM module wants to find out :wink:) -- for this case passing a memfd works well, but it doesn't work for the sudo case.

But ideally I'd like to use *one* communication mechanism of identification for both cases. One idea that I have about that is that the PAM module can check its parent PID. If it's cockpit-ws (which spawns cockpit-session), then it can ask cockpit-tls (via its Unix socket) about "give me the certificate for this instance of cockpit-ws", which cockpit-tls knows about directly. In the sudo case, the parent process will be something else (cockpit-bridge?), but then it has the session's uid/user name. So cockpit-tls needs a way to map a user name to a (certificate, cockpit-ws instance) pair, but that doesn't seem to be too difficult -- cockpit-session could just tell cockpit-ws about that, through the Unix socket or through a file in /run.

AFAICS this only involves direct communication between the trusted cockpit-ws and the cockpit_pam_cert module. cockpit-ws can't inject untrustworthy information -- as long as it stays the same process, its pid will remain, and if it crashes itself it can only cause a DoS. But we can't avoid this DoS scenario anyway.

This also means that we don't need fancy stuff like memfds, which makes this applicable to RHEL 7. It also doesn't use anything which isn't containerizable (like kernel keyrings), so all of that should work just fine in the cockpit/ws container.

Reply:
Further simplification: It doesn't even need to be an Unix socket -- cockpit-tls can just export trusted certificates as files in /run/, and the PAM module can validate their ownership.